goinp/goinp.go

```diff
@@ -120,37 +120,47 @@ func AskOptions(title string, defaultValue string, optional bool, options ...str
 	}
 
 	for i, option := range options {
-		log.Printf("[%d] : %s", i+1, option)
+		fmt.Printf("[%d] : %s\n", i+1, option)
 	}
 
 	for {
 		fmt.Print("Type in the option's number, then hit Enter: ")
 
-		answer, err := bufio.NewReader(os.Stdin).ReadString('\n')
+		answer, err := reader.ReadString('\n')
 		if err != nil {
-			log.Errorf("failed to read input value")
+			fmt.Printf(colorstring.Red("failed to read input value") + "\n")
 			continue
 		}
 
 		optionNo, err := strconv.Atoi(strings.TrimSpace(answer))
 		if err != nil {
-			log.Errorf("failed to parse option number, pick a number from 1-%d", len(options))
+			fmt.Print(colorstring.Red(fmt.Sprintf("failed to parse option number, pick a number from 1-%d", len(options))) + "\n")
 			continue
 		}
 
 		if optionNo-1 < 0 || optionNo-1 >= len(options) {
-			log.Errorf("invalid option number, pick a number 1-%d", len(options))
+			fmt.Print(colorstring.Red(fmt.Sprintf("invalid option number, pick a number 1-%d", len(options))) + "\n")
 			continue
 		}
 
-		if options[optionNo-1] == customValueOptionText {
-			return AskOptions(title, "", true)
+		if optionNo == len(options) && optional {
+			return askOptionsWithReader(title, "", true, reader)
 		}
 
 		return options[optionNo-1], nil
 	}
 }
 
+// AskOptions ...
+func AskOptions(title string, defaultValue string, optional bool, options ...string) (string, error) {
```

Comment:
why do not you wire in the reader as a function input?

that way you would not need to override the os.Stdin and out in the tests and I think you could use the log package for logging.

Reply:
This answer proves only another solution's possibility, not this solution's correctness. Please fill in your references that are states that this method is agains anything.

Reply:
- Writing to terminal input buffer is only possible via os.Stdin (or to /dev/stdin)
- Cannot convert a io.Reader back to *os.File
- Test environment has a lack of file descriptors so os.Stdin is nit set by default (not terminal session), and it is a common thing to set os.stdin and then reading the written stuff to validate that the method could, and wrote the expected values into it.